File: frontend/banque-sang-ui/src/app/core/services/stock.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Sang, StockGroupe } from '../../shared/models/stock.model';

@Injectable({
  providedIn: 'root'
})
export class StockService {
  private readonly apiUrl = `${environment.apiUrl}/stock`;

  constructor(private readonly http: HttpClient) {}

  /**
   * Récupère toutes les poches de sang.
   */
  getAll(): Observable<Sang[]> {
    return this.http.get<Sang[]>(this.apiUrl);
  }

  /**
   * Récupère les statistiques de stock groupées par groupe sanguin.
   */
  getStockParGroupe(): Observable<StockGroupe[]> {
    return this.http.get<StockGroupe[]>(`${this.apiUrl}/groupes`);
  }

  /**
   * Récupère les poches proches de la péremption ou déjà périmées.
   */
  getAlertes(): Observable<Sang[]> {
    return this.http.get<Sang[]>(`${this.apiUrl}/alertes`);
  }
}

File: backend/BanqueSang/BanqueSang.Core/Services/StockService.cs
using BanqueSang.Core.Entities;
using BanqueSang.Core.Interfaces;
using BanqueSang.Core.Models;

namespace BanqueSang.Core.Services;

public class StockService : IStockService
{
    private readonly ISangRepository _sangRepository;

    public StockService(ISangRepository sangRepository)
    {
        _sangRepository = sangRepository;
    }

    /// <summary>
    /// Récupère toutes les poches de sang.
    /// </summary>
    public async Task<IEnumerable<Sang>> GetAllAsync()
    {
        return await _sangRepository.GetAllAsync();
    }

    /// <summary>
    /// Récupère les statistiques de stock par groupe sanguin.
    /// </summary>
    public async Task<IEnumerable<StockGroupeResult>> GetStockParGroupeAsync()
    {
        return await _sangRepository.GetStockParGroupeAsync();
    }

    /// <summary>
    /// Récupère les poches périmées ou proches de leur date de péremption.
    /// </summary>
    public async Task<IEnumerable<Sang>> GetAlertesAsync()
    {
        return await _sangRepository.GetAlertesAsync();
    }
}